Client/Frontend/Browser/Tabs/RemoteTabs/RemoteTabsPanel.swift

```diff
@@ -0,0 +1,379 @@
+            showEmptyTabsViewWith(.noClients)
+            return
+        }
+
+        let nonEmptyClientAndTabs = clientAndTabs.filter { !$0.tabs.isEmpty }
+        guard !nonEmptyClientAndTabs.isEmpty else {
+            showEmptyTabsViewWith(.noTabs)
+            return
+        }
+
+        let tabsPanelDataSource = RemoteTabsClientAndTabsDataSource(actionDelegate: remoteTabsPanel,
+                                                                    clientAndTabs: nonEmptyClientAndTabs,
+                                                                    theme: themeManager.currentTheme)
+        tabsPanelDataSource.collapsibleSectionDelegate = self
+        tableViewDelegate = tabsPanelDataSource
+
+        tableView.reloadData()
+    }
+
+    func refreshTabs(state: RemoteTabsPanelState, updateCache: Bool = false, completion: (() -> Void)? = nil) {
+        ensureMainThread { [self] in
+            self.state = state
+            performRefresh(updateCache: updateCache, completion: completion)
+        }
+    }
+
+    private func performRefresh(updateCache: Bool, completion: (() -> Void)?) {
+        // Short circuit if the user is not logged in
+        guard state.allowsRefresh else {
+            endRefreshing()
+            showEmptyTabsViewWith(.notLoggedIn)
+            return
+        }
+
+        // TODO: Send Redux action to get cached clients & tabs, update once new state is received. Forthcoming.
+        // store.dispatch(RemoteTabsPanelAction.refreshCachedTabs)
+
+        // Previous code for reference. Will be removed soon.
+        /*
+        // Get cached tabs.
+        profile.getCachedClientsAndTabs().uponQueue(.main) { [weak self] result in
+            guard let clientAndTabs = result.successValue else {
+                self?.endRefreshing()
+                self?.showEmptyTabsViewWith(.failedToSync)
+                return
+            }
+
+            // Update UI with cached data.
+            self?.updateDelegateClientAndTabData(clientAndTabs)
+
+            if updateCache {
+                
+                // TODO: Send Redux action to refresh tabs, and then update once new state is received. Forthcoming.
+                // store.dispatch(RemoteTabsPanelAction.refreshTabs)
+                
+                // Original code:
+                // self?.profile.getClientsAndTabs().uponQueue(.main) { result in
+                //     if let clientAndTabs = result.successValue {
+                //         // Update UI with updated tabs.
+                //         self?.updateDelegateClientAndTabData(clientAndTabs)
+                //     }
+                // 
+                //     self?.endRefreshing()
+                //     completion?()
+                // }
+            } else {
+                self?.endRefreshing()
+                completion?()
+            }
+        }
+         */
+    }
+
+    private func showEmptyTabsViewWith(_ error: RemoteTabsErrorDataSource.ErrorType) {
```

Comment:
This one is a matter of preference but I always found it weird that we have 2 data sources one for the tableview and another for the error states I don't think this is the standard usage of data sources. I think we should move away from it in the refactor unless there is an excellent reason to keep this logic. I think it is more iOS-app like and how we use in the rest of the project to have a single data source for the table view and have a ErrorView with its own state and we show either the table view or the error view

Reply:
> I always found it weird that we have 2 data sources one for the tableview and another for the error states I don't think this is the standard usage of data sources.

Yeah this is just following the existing pattern in the codebase. There seem to be a lot of rules for things like this that I'm not aware of, but they also aren't documented in the tickets (AFAICS) so I don't know in advance how we want to refactor some of these things.

Do we have any specs or up-to-date docs that list these various patterns / style rules that I should be following? Thank you @yoanarios @OrlaM

Reply:
@yoanarios I've been digging into this and I think it's going to be a bit messy to get rid of these data sources right now, though I agree we should refactor them eventually. But there are 5 different error states that the `RemoteTabsErrorDataSource` (and related cell) handles, and separate logic within the clients-and-tabs data source, and some of that is difficult to exercise in the newer VCs without more of the Redux wiring in place. I'm concerned refactoring all of that at this stage is going to introduce a lot of easy-to-miss bugs.

I think what might be better is for me to add a separate task (to the tab tray epic) to refactor away those two data sources and revisit it once we're further along with the wiring. We need those data sources for the legacy VCs anyway, so they have to be maintained at least for a little while longer.

Let me know if that ☝️ plan works for you.

Reply:
@yoanarios I think (per Slack chat) all feedback here should now be addressed.